### PR TITLE
Karatsuba multiplication & parameter test sets

### PR DESCRIPTION
**Pull request by ewynx**
tldr; Karatsuba multiplication and test parameters :D 

This adds Karatsuba multiplication for lengths 13, 17, 18, 26 & 34. There are general functions to easily add more level-1 Karatsuba multiplications. For level-2 Karatsuba you need to copy the code of another Karatsuba-2 multiplication and adjust the params.

I only replaced the schoolbook multiplication in `evaluate_quadratic_expression`, because this is a constrained function.  

The main reason I haven't added Karatsuba for the other lengths is because it is probably best to check the benchmark while doing it, to determine whether it should be 1 or 2 level, or even schoolbook. Since the whole idea is to be able to pick the fastest multiplication for a certain array length, it makes sense to add a karatsuba multiplication after doing the benchmarking. For benchmarking I used https://github.com/hashcloak/noir-bigint-bench

### CODE REVIEW

File: resources/Create_params.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    fn modulus_bits() -> u64;
+
+    fn mult(a: [Field; N], b: [Field; N]) -> ArrayX<Field, N, 2>;
+}
+```
+
+## N
+
+`N` is the number of limbs needed to represent the bigint in radix-120, with enough space to represent `2p` and `redc` which is needed for next steps: `ceil(modulus_bits+1 / 120)`.
```
Comment (hdvanegasm) [inline resources/Create_params.md]: I was thinking a long time about saying "radix-120" or "120-bit radix". I think the last one is better because when I read "radix-120" I feel that the basis of the representation is 120 and not 2^120. However, this is a minor thing that we can discuss later.

File: resources/Create_params.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+Return modulus `p` in radix-120 in little-endian. 
+
+## double_modulus
+
+Return `2p` in radix-120 in little-endian.
+
+## redc_param, k
+
+These parameters are used for Barrett reduction: `redc = floor(2^2k / p)` for `k = modulus_bits` or `k = modulus_bits+1`.
```
Comment (hdvanegasm) [inline resources/Create_params.md]: According to Zac, k should be `modulus_bits()` for the general case.

File: resources/parigp_scripts.md [hunk 3]
```diff
@@ -1,4 +1,52 @@
...
+# PariGP scripts to support testing/usage
+
+## Generating parameters for BigNumParamsTrait
+
+### 1. Generate random modulus for bitlength
+
+For example, generate a random modulus `p` of 256 bits. 
+```
+bits = 3000;
+random(2^bits)
```
Comment (hdvanegasm) [inline resources/parigp_scripts.md]: I know that this was to generate something with 3000 bits, but eventualy we should modify this to match with the description.

File: src/mul.nr [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+  **/
+fn subtract_arrays<N>(a: ArrayX<Field, N, 2>, b: ArrayX<Field, N, 2>, c: ArrayX<Field, N, 2>) -> ArrayX<Field, N, 2> {
+    let mut res = ArrayX { segments: [[0;N], [0; N]] };
+    for i in 0..2 {
+        for j in 0..N {
+            res.segments[i][j] = a.segments[i][j] - b.segments[i][j] - c.segments[i][j]; 
+        }
+    }
+    res
+}
```
Comment (hdvanegasm) [inline src/mul.nr:49]: Possible improvement: we can use methods `set()` and `get()` to make the code more readable. However, I'm not sure if we can do `N + N` for a generic parameter `N`.

File: src/mul.nr [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+
+/**
+  * @brief For karatsuba multiplication, returns values r0, r1, r2
+  *        which represent the lower, (overlapping) middle and higher limbs of the result
+  * @description: this computes for a, b, split in the middle as a = [a_lo, a_hi], b = [b_lo, b_hi]:
+  *               (If the length is odd, the lower half is 1 smaller than the higher half)
+  *               r0 = a_lo * b_lo
+  *               r2 = a_hi * b_hi
+  *               r1 = a_lo*b_hi + a_hi*b_lo
```
Comment (hdvanegasm) [inline src/mul.nr:58]: Minor comment: we should add that the computation of `r1` is `r1 = (a_lo + a_hi) * (b_lo + b_hi) - r2 - r0` here in the docs.

File: src/mul.nr [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+
+    // The following is calculated:
+    // r0 = a_lo * b_lo
+    // r2 = a_hi * b_hi
+    // r1 = (a_lo+a_hi)(b_lo+b_hi) - r0 - r2
+    //    = a_lo*b_lo + a_lo*b_hi + a_hi*b_lo + a_hi*b_hi - a_lo*b_lo - a_hi*b_hi
+    //    = a_lo*b_hi + a_hi*b_lo
+
+    // a_lo, b_lo, a_hi, b_hi have (M of M-1) limbs of 120 bits
+    // r0, r2 have 2M-1 limbs of 240+log(2M-1) bits
```
Comment (hdvanegasm) [inline src/mul.nr:96]: Minor comment: I think that `r1` has `2M - 1` limbs of at most 242 + log_2(M) bits, because we are multiplying two numbers with M limbs. Correct this also in the below comments.

File: src/mul.nr [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+    let r2: ArrayX<Field, M, 2> = schoolbook_mult(a_hi, b_hi);
+
+    // added_a, added_b have (M of M-1) limbs of 121 bits
+    let added_a = add_arrays(a_lo, a_hi);
+    let added_b = add_arrays(b_lo, b_hi);
+    // r1 has 2M-1 limbs of 242+log(2M-1) bits
+    let mut r1: ArrayX<Field, M, 2> = schoolbook_mult(added_a, added_b);
+    // This subtraction gives no issues if used in combination with combining these results as follows:
+    // [r0_lo, r0_hi + r1_lo, r1_hi + r2_lo, r2_hi]
+    // because all limbs will be positive, even if in intermediate value r1 there may be "negative" values
```
Comment (hdvanegasm) [inline src/mul.nr:107]: I don't fully understand this comment. Why isn't it an overflow there? We can discuss this in a meeting

File: src/lib.nr [hunk 8]
```diff
@@ -30,22 +33,31 @@ trait BigNumParamsTrait<N> {
...
+
     /**
      * @brief modulus_bits = log2(modulus) rounded up
      **/
     fn modulus_bits() -> u64;
+
+    /**
+     * @brief efficient multiplication of 2 N-length arrays. To use in constrained fn
+     **/
+    fn mult(a: [Field; N], b: [Field; N]) -> ArrayX<Field, N, 2>;
```
Comment (hdvanegasm) [inline src/lib.nr:55]: I think that the trait `BigNumParams` should store just parameters for the big-integer, however `mult()` is not itself a param. I suggest to move this to another struct or trait like `BigNum`. What do you think?

**Review by hdvanegasm — APPROVED** [review]
I have left informational comments. I don't see any issue related to logic. We can merge.

File: resources/Create_params.md (hunk 1 above)
Comment (ewynx) [reply resources/Create_params.md]: you are right, this should be radix-2^120, because the base is 2^120

File: resources/Create_params.md (hunk 2 above)
Comment (ewynx) [reply resources/Create_params.md]: removed k param and replaced it with modulus_bits

File: src/mul.nr (hunk 4 above)
Comment (ewynx) [reply src/mul.nr:49]: this is actually a possible improvement the other way around I was thinking about. I think accessing `segments[i][j]` might be faster than `segments.get(i)`, because the latter calculates the value of `i` and `j` which uses a division and a mod. So I think we can try to replace and if it makes any difference. I agree with you that if they have the same gatecount, `get()` and `set()` is more readable.

File: src/mul.nr (hunk 7 above)
Comment (ewynx) [reply src/mul.nr:107]: Discussed

File: src/mul.nr (hunk 5 above)
Comment (ewynx) [reply src/mul.nr:58]: those comments are on line 91 and onwards, because those are implementation details I think that is sufficient.

File: src/lib.nr (hunk 8 above)
Comment (ewynx) [reply src/lib.nr:55]: hmm good point.. however then it might be a bit more cumbersome for a user because they would have to implement 2 separate traits. Maybe a renaming would be more user friendly?

Comment (ewynx) [reply src/lib.nr:55]: we'll adjust this later
